src/odoc/occurrences.ml

```diff
@@ -0,0 +1,203 @@
+      | `Exception (parent, _) -> do_ parent
+      | `ExtensionDecl (parent, _, _) -> do_ parent
+      | `Class (parent, _) -> do_ parent
+      | `Value (parent, _) -> do_ parent
+      | `ClassType (parent, _) -> do_ parent
+      | `Root _ -> incr tbl id
+      | `SourcePage _ | `Page _ | `LeafPage _ | `SourceLocation _
+      | `CoreException _ | `Label _ | `SourceLocationMod _ | `Result _
+      | `AssetFile _ | `SourceDir _ | `SourceLocationInternal _ ->
+          assert false
+    in
+    let _htbl = add ~kind:`Direct id in
+    ()
+
+  let rec get t id =
+    let ( >>= ) = Option.bind in
+    let do_ parent =
+      get t (parent :> key) >>= fun { sub; _ } -> H.find_opt sub id
+    in
+    match id.iv with
+    | `InstanceVariable (parent, _) -> do_ parent
+    | `Parameter (parent, _) -> do_ parent
+    | `Module (parent, _) -> do_ parent
+    | `ModuleType (parent, _) -> do_ parent
+    | `Method (parent, _) -> do_ parent
+    | `Field (parent, _) -> do_ parent
+    | `Extension (parent, _) -> do_ parent
+    | `ExtensionDecl (parent, _, _) -> do_ parent
+    | `Type (parent, _) -> do_ parent
+    | `Constructor (parent, _) -> do_ parent
+    | `Exception (parent, _) -> do_ parent
+    | `Class (parent, _) -> do_ parent
+    | `Value (parent, _) -> do_ parent
+    | `ClassType (parent, _) -> do_ parent
+    | `Root _ -> H.find_opt t id
+    | `SourcePage _ | `Page _ | `LeafPage _ | `CoreType _ | `SourceLocation _
+    | `CoreException _ | `Label _ | `SourceLocationMod _ | `Result _
+    | `AssetFile _ | `SourceDir _ | `SourceLocationInternal _ ->
+        assert false
+
+  let rec iter f tbl =
+    H.iter
+      (fun id v ->
+        iter f v.sub;
+        f id v)
+      tbl
+end
+
+let count ~dst ~warnings_options:_ directories include_hidden =
+  let htbl = H.create 100 in
+  let f () (unit : Odoc_model.Lang.Compilation_unit.t) =
+    let incr tbl p =
+      let p = (p :> Odoc_model.Paths.Path.Resolved.t) in
+      let id = Odoc_model.Paths.Path.Resolved.identifier p in
+      if (not (Odoc_model.Paths.Path.Resolved.is_hidden p)) || include_hidden
+      then Occtbl.add tbl id
+    in
+    let () =
+      List.iter
+        (function
+          | ( Odoc_model.Lang.Source_info.Module
+                { documentation = Some (`Resolved p); _ },
+              _ ) ->
+              incr htbl p
+          | Value { documentation = Some (`Resolved p); _ }, _ -> incr htbl p
+          | ClassType { documentation = Some (`Resolved p); _ }, _ ->
+              incr htbl p
+          | ModuleType { documentation = Some (`Resolved p); _ }, _ ->
+              incr htbl p
+          | Type { documentation = Some (`Resolved p); _ }, _ -> incr htbl p
+          | _ -> ())
+        (match unit.source_info with None -> [] | Some i -> i.infos)
+    in
+    ()
+  in
+  fold_dirs ~dirs:directories ~f ~init:() >>= fun () ->
+  Fs.Directory.mkdir_p (Fs.File.dirname dst);
+  let oc = open_out_bin (Fs.File.to_string dst) in
```

Comment:
Note for later: It seems that all Odoc commands create the parent directories of files they create. It would be nice to have this written down in the form of a factorized function.